[BUG/FEATURE] Fix Sym=False,  new checkpoint_format = gptq_v2

Qubitium opened the pull request:
@qwopqwop200  This is the rebase of your PR at #559 with some modifications. Should be ready soon after we verify quantize, inference, and add some tests. 

### Reason For PR:

`sym=False` was practically unusable due to post-quantization avg_loss per layer/PPL vs `sym=True` . @qwopqwop200 fixed the bad/suboptimal math. Now `sym=False` will most likely match or decrease avg_loss/layer vs `sym=True` and improve post-quant PPL for many models. 

### Core Changes:

1. Rebase PR #559 with `main`: allow usable `sym=False` quantization and use `checkpoint_format=gptq_v2` to store new checkpoint format. Compat runtime dynamic convert of all `checkpoint_format=gptq` to `gptq_v2` on load. 

### Misc Changes not directly related to sym=False code:

4. Complete Todo: Use accelerate 0.29.2 to load checkpoint. 
5. Consistency: Move cohere/starcoder2 to 4.39.0 release check
6. Usability: Catch and alert user on how to fix quant/torch error caused by low damp/nsamples #572 as this happens much more frequently than I had expected. Ran into 2-3 instances of this error on multiple models during testing for this PR when using low `nsamples`  + low `damp=0.005` to speed up quants. 
7. Simplify: optimized packing regression alert message/code to user
8. Feature: Quant Stat Log 1/2:  store per layer quant stats (`layer #`, `module name`, `avg loss`, `duration`)  in dict/slice and return to user va `quant_log = model.quantize()`
9. Feature: Quant Stat Log 2/2:  pass saved quant_log to `quantize(quant_log=saved_quant_log)` to generate auto-avg_loss diff in progress. Sample diff output in later messages of this discussion. 
10. Usability: Use tqdm for layer loop so users can have an estimate of quant remaining time. 

TODO:

* [x] Add `sym=False` tests
* [x] Validate and fix failing tests
* [x] Failed: Check if third party vllm/sglang needs kernel modification for the new `gptq_v2` format
* [x] Check if third party vllm/sglang needs kernel modification for the new `gptq` (v1) format using `sym=False` in this PR


PASSING TESTS:
* [x] Compat: vllm/sglang compatible  with gptq (v1) generated with this PR
* [x] Compat load of `checkpoint_format=gptq`(v1)
* [x] `sym=False` consistently generate lower `avg_loss` than `sym=True`
* [x] Regression Test: `sym=True` in PR generates same math/`avg_loss` for layers as `sym=True` in `main`
* [x] `test_serialization.py` 
* [x] `test_quantization.py`
* [x] `test_shared_loading.py`
* [x] `test_awq_compatibility_generation.py`  (note: awq cache generated by `main` is not compatible with pr. fixed with version file name adding `v2`)
* [x] `test_q4.py`

FAILING TESTS:
* [x] compat: vllm/sglang not compatible gptq_v2
* [x] `test_triton.py` (never got this to work on main)
* [x] `test_repacking.py` (never got this to work on main)  

Original PR #559 notes duplicated here for ref:

> * [x]  check work marlin
> * [x]  check work exllama
> * [x]  check work exllama2
> * [x]  check work qigen
> * [x]  check work triton
> * [x]  check work cuda
> * [x]  ~check work cuda old~(There is a bug in autogptq's main unrelated to this pr, so it cannot be confirmed.)
> * [x]  check work cuda pytorch
> * [x]  check work cuda old pytorch
> * [x]  check support old version save
> * [x]  check support old version load
> * [x]  check support new version save
> * [x]  check support new version load
> 
> I am removing this line because it is not only computationally unnecessary, but also makes sym=False impossible. However, it breaks backwards compatibility, so I making it the default to use the old save format.
> 
> ## related pr
> #354
> #325
> I removed the following line: https://github.com/AutoGPTQ/AutoGPTQ/blob/main/auto_gptq/nn_modules/qlinear/qlinear_cuda.py#L169 This is an unnecessary line. And this line makes it impossible to sym='False' If sym='False' I get a reduction in ppl.
> 
> opt-125m(act-order)	Bits	group-size	Wikitext2
> sym=True	4	128	29.875
> sym=False	4	128	29.221
> llama2(act-order)	Bits	group-size	Wikitext2
> sym=True	4	128	5.254
> sym=False	4	128	5.214

Qubitium commented:
Misc Changes (wip notes) not directly related to sym=False code:

1. Safetensor forced on `save_pretrain()` + `save_quantized()`. removed `.bin` option. 
2. Use accelerate 0.29.2 to load checkpoint. Remove todo
3. Move cohere/starcoder2 to 4.39.0 release check
4. Catch and alert user on how to fix quant/torch error caused by low damp/nsamples 

TODO:

1. Add tests
2. Validate and fix failing tests
3. Check if third party vllm/sglang needs kernel modification for the new `gptq_v2` format

Qubitium commented:
### TinyLlama 1.1B sym=True

```python 
+---------+------------------+------------+----------+-----------------+-------------+
|   layer | name             |   avg_loss |     time |   diff_avg_loss |   diff_time |
+=========+==================+============+==========+=================+=============+
|       1 | self_attn.k_proj |   0.006130 | 0.585682 |        0.003079 |   -0.599414 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.v_proj |   0.000175 | 0.463799 |        0.000033 |   -0.609846 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.q_proj |   0.010899 | 0.471033 |        0.005346 |   -0.425677 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.o_proj |   0.000005 | 0.470546 |        0.000001 |   -0.652694 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.up_proj      |   0.005891 | 0.471234 |        0.001170 |   -0.642720 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.gate_proj    |   0.006253 | 0.473800 |        0.001450 |   -0.630586 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.down_proj    |   0.000042 | 1.307003 |        0.000009 |   -1.763714 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | self_attn.k_proj |   0.029761 | 0.470537 |        0.011130 |   -0.630198 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | self_attn.v_proj |   0.001209 | 0.467803 |        0.000209 |   -0.617392 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | self_attn.q_proj |   0.067034 | 0.470898 |        0.020041 |   -0.430467 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | self_attn.o_proj |   0.000054 | 0.464447 |        0.000011 |   -0.616824 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | mlp.up_proj      |   0.013213 | 0.473601 |        0.002401 |   -0.613731 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | mlp.gate_proj    |   0.014810 | 0.469599 |        0.002817 |   -0.609236 |
+---------+------------------+------------+----------+-----------------+-------------+
|       2 | mlp.down_proj    |   0.000067 | 1.304770 |        0.000014 |   -1.699867 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | self_attn.k_proj |   0.027362 | 0.465101 |        0.007338 |   -0.615543 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | self_attn.v_proj |   0.000914 | 0.465002 |        0.000160 |   -0.613301 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | self_attn.q_proj |   0.057088 | 0.468296 |        0.011540 |   -0.478751 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | self_attn.o_proj |   0.000079 | 0.468615 |        0.000015 |   -0.612965 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | mlp.up_proj      |   0.024809 | 0.470201 |        0.004419 |   -0.614573 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | mlp.gate_proj    |   0.027832 | 0.471135 |        0.004975 |   -0.604681 |
+---------+------------------+------------+----------+-----------------+-------------+
|       3 | mlp.down_proj    |   2.708260 | 1.299222 |        1.965734 |   -1.705520 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | self_attn.k_proj |   0.036803 | 0.469426 |        0.006950 |   -0.608829 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | self_attn.v_proj |   0.001792 | 0.464624 |        0.000247 |   -0.614279 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | self_attn.q_proj |   0.085306 | 0.469892 |        0.015617 |   -0.479200 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | self_attn.o_proj |   0.000033 | 0.466727 |        0.000008 |   -0.613243 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | mlp.up_proj      |   0.034699 | 0.469619 |        0.005476 |   -0.615357 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | mlp.gate_proj    |   0.040107 | 0.467804 |        0.006557 |   -0.611297 |
+---------+------------------+------------+----------+-----------------+-------------+
|       4 | mlp.down_proj    |   0.000104 | 1.309098 |        0.000017 |   -1.697408 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | self_attn.k_proj |   0.048844 | 0.468561 |        0.008955 |   -0.608376 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | self_attn.v_proj |   0.003229 | 0.463883 |        0.000639 |   -0.612935 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | self_attn.q_proj |   0.123073 | 0.469848 |        0.024918 |   -0.488955 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | self_attn.o_proj |   0.000151 | 0.467359 |        0.000032 |   -0.611210 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | mlp.up_proj      |   0.076759 | 0.470483 |        0.014432 |   -0.612712 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | mlp.gate_proj    |   0.112788 | 0.468013 |        0.021772 |   -0.605751 |
+---------+------------------+------------+----------+-----------------+-------------+
|      10 | mlp.down_proj    |   0.000636 | 1.312896 |        0.000129 |   -1.661242 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.k_proj |   0.057187 | 0.461225 |        0.012757 |   -0.069699 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.v_proj |   0.021670 | 0.461217 |        0.003897 |   -0.080796 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.q_proj |   0.170942 | 0.463462 |        0.044763 |   -0.073541 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.o_proj |   0.007679 | 0.468042 |        0.001327 |   -0.068797 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.up_proj      |   0.324943 | 0.464848 |        0.071946 |   -0.067675 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.gate_proj    |   0.516380 | 0.461897 |        0.114707 |   -0.074716 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.down_proj    |   0.653858 | 1.284941 |        0.157109 |   -0.211105 |
+---------+------------------+------------+----------+-----------------+-------------+
```

Qubitium commented:
TinyLlama 1.1B `sym=True` + `checkpoint_format=gptq_v2` of PR (diff is vs `sym=True` of `main`)

Regression test passed. There is no math/`avg_loss` diff between PR and main when `sym=True`. 

```python
+---------+------------------+------------+----------+-----------------+-------------+
|   layer | name             |   avg_loss |     time |   diff_avg_loss |   diff_time |
+=========+==================+============+==========+=================+=============+
|       1 | self_attn.k_proj |   0.406458 | 0.628311 |        0.000000 |    0.000592 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.v_proj |   0.002031 | 0.522467 |        0.000000 |    0.006760 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.q_proj |   0.503623 | 0.514812 |        0.000000 |    0.005269 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | self_attn.o_proj |   0.000061 | 0.522506 |        0.000000 |    0.020049 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.up_proj      |   0.108624 | 0.522778 |        0.000000 |    0.008349 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.gate_proj    |   0.115655 | 0.520952 |        0.000000 |    0.013677 |
+---------+------------------+------------+----------+-----------------+-------------+
|       1 | mlp.down_proj    |   0.000160 | 1.442813 |        0.000000 |    0.029875 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | self_attn.k_proj |   1.037582 | 0.513118 |        0.000000 |    0.014547 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | self_attn.v_proj |   0.064296 | 0.510034 |        0.000000 |    0.011178 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | self_attn.q_proj |   2.442628 | 0.510062 |        0.000000 |    0.012037 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | self_attn.o_proj |   0.005665 | 0.513529 |        0.000000 |    0.013509 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | mlp.up_proj      |   1.633932 | 0.519150 |        0.000000 |    0.015929 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | mlp.gate_proj    |   2.260344 | 0.515065 |        0.000000 |    0.010460 |
+---------+------------------+------------+----------+-----------------+-------------+
|      11 | mlp.down_proj    |   0.012365 | 1.441533 |        0.000000 |    0.050790 |
+---------+------------------+------------+----------+-----------------+---------
|      21 | self_attn.k_proj |   1.126530 | 0.511904 |        0.000000 |    0.006600 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | self_attn.v_proj |   0.337632 | 0.512384 |        0.000000 |    0.004558 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | self_attn.q_proj |   3.222927 | 0.514188 |        0.000000 |    0.004995 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | self_attn.o_proj |   0.056764 | 0.515278 |        0.000000 |    0.005255 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | mlp.up_proj      |   6.612084 | 0.518280 |        0.000000 |    0.000549 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | mlp.gate_proj    |   8.599123 | 0.514044 |        0.000000 |    0.000147 |
+---------+------------------+------------+----------+-----------------+-------------+
|      21 | mlp.down_proj    |   0.427881 | 1.425859 |        0.000000 |    0.011282 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.k_proj |   1.190292 | 0.512079 |        0.000000 |    0.002728 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.v_proj |   0.454321 | 0.512852 |        0.000000 |    0.006104 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.q_proj |   3.392396 | 0.512543 |        0.000000 |    0.000779 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | self_attn.o_proj |   0.098364 | 0.521686 |        0.000000 |    0.010890 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.up_proj      |   7.234830 | 0.516273 |        0.000000 |    0.000210 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.gate_proj    |  11.574709 | 0.510233 |        0.000000 |    0.000868 |
+---------+------------------+------------+----------+-----------------+-------------+
|      22 | mlp.down_proj    |   2.122912 | 1.403076 |        0.000000 |   -0.021529 |
+---------+------------------+------------+----------+-----------------+-------------+
```

Qubitium commented:
@qwopqwop200  The cuda old tests was not passing was likely caused by caching of auto-marlin conversion. It also failed for me in `test_q4` and magically disappeared when I updated the cache file name so caches from pre-pr is never loaded. 

@fxmarty  The caching of dyamic awq/marlin processing is becoming a more of a problem. I ran into 2 more cases of the caching causing silent fails: 1) test_q4 old cuda (same problem as @qwopqwop200) and awq compat test failing (also caused by incompatible cache file).

Qubitium commented:
@qwopqwop200  Right now the only remaining tests failing is test_q4 triton. Setting `use_tritonv2=False` makes the tests pass.  I  do not have cuda/triton skills to find the cause this issue. Need your assistance to track down this problem. Thanks.

```
# run triton test only
pytest test_q4.py::TestsQ4Triton
```

```python
FAILED test_q4.py::TestsQ4Triton::test_generation_no_act_order - AssertionError: '<s> I am in Paris and<unk><unk><unk><unk><unk><unk><unk><unk><[255 chars]unk>' != '<s> I am in Paris and I am going to the Louvre Museum. What ti[161 chars] get'
FAILED test_q4.py::TestsQ4Triton::test_generation_with_act_order - AssertionError: '<s> I am in Paris and<unk><unk><unk><unk><unk><unk><unk><unk><[255 chars]unk>' != '<s> I am in Paris and it is a beautiful day. I am sitting in a[203 chars] and'
```

qwopqwop200 commented:
remove this line https://github.com/AutoGPTQ/AutoGPTQ/blob/main/auto_gptq/nn_modules/qlinear/qlinear_tritonv2.py#L146
and this line
https://github.com/AutoGPTQ/AutoGPTQ/blob/main/auto_gptq/nn_modules/triton_utils/dequant.py#L79

I think this will probably work.

Qubitium commented:
@qwopqwop200  Thank you! triton tests passed. I will re-run full `test_q4`.

liuy-2019 commented:
If bit=8, it has this bug? I see all the changes is about 4 bit, what about 8 bit?

Qubitium commented:
> If bit=8, it has this bug? I see all the changes is about 4 bit, what about 8 bit?

This PR will affect all bits that gptq supports. [ 2, 4, 8 ]. The tests only validates that there are no errors with 4 bit quants since that is the most popular bit version for balance of speed and quality. You can compare the PPL and/or avg-loss values between this PR and main on 8bit as well.

fxmarty commented:
@qwopqwop200 @Qubitium thanks a lot for fixing this. I will review & merge hopefully today/in the coming days, and do a release thereafter.

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -351,13 +360,45 @@ def tmp(_, inp, out):
+                            static_groups=self.quantize_config.static_groups,
+                        )
+
+                        stat = {"layer": i + 1, "module": name, "avg_loss": avg_loss, "time": duration}
```
```suggestion
                        stat = {"layer": i, "module": name, "avg_loss": avg_loss, "time": duration}
```

Qubitium replied:
@fxmarty  Can we stick with  `i + 1` so tqdm and layer logs are consistent? using `i + 1` and then bam, having the stat logs printin middle of the log as `i`  and at the very end (summary) is jarring. Also, the final quant_log table is meant for visual consumption more so than code consumption.

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -517,7 +562,6 @@ def push_to_hub(
     def save_quantized(
         self,
         save_dir: str,
-        use_safetensors: bool = True,
```
Why removing it & how is it related to this PR?

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -865,6 +929,7 @@ def from_quantized(
         if use_safetensors:
             extensions.append(".safetensors")
         else:
+            logger.warning("Deprecation: Loading of non-safetensor weights will be removed in the future. Please convert your weights to safetensor format. ")
```
Why would it be?

Qubitium replied:
Reason for the change is that I don't believe there is a non-safetensor future based on this: 

https://github.com/huggingface/transformers/pull/29846/files
and comment/reason behind the PR here: https://github.com/huggingface/transformers/issues/29500#issuecomment-1983289037

> This is to continue pushing safetensors forward as an alternative to raw .bin checkpoints as we see more bad actors with pure .bin checkpoints. Thanks again for your report!

HF is taking very aggressive steps to speed up safetensor adoption due to threat-actors using .bin files. I am not privy to hf security stats but this opened my eyeball as forcing safetensor usage in the long run.

fxmarty replied:
Let's discuss this in an other PR/issue, so as to keep this one light if possible.

Although safetensors is more widely adopted (and the default in HF libs), for now most (all?) libraries still support .bin recommended by PyTorch, so let's support as well (although not really tested in AutoGPTQ tests).

AFAIK there is no plan to support safetensors in PyTorch natively

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -1130,7 +1195,7 @@ def skip(*args, **kwargs):
                     inject_fused_mlp = False
 
-            load_checkpoint_in_model(
+            accelerate.utils.modeling.load_checkpoint_in_model(
```
good

fxmarty commented on auto_gptq/modeling/_utils.py:
```diff
@@ -146,6 +147,51 @@ def make_quant(
+    qlinear_kernel: nn.Module,
+):
+    use_qigen = qlinear_kernel.QUANT_TYPE == "qigen"
+
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, qlinear_kernel):
+            if v2_format:
+                if use_qigen:
+                    submodule.zeros.data += 1
+                else:
+                    if quantize_config.bits == 2:
+                        submodule.qzeros.data += 0b01010101010101010101010101010101
+                    elif quantize_config.bits == 3:
+                        submodule.qzeros.data[:,range(0,submodule.qzeros.data.shape[1],3)] += 0b00100100100100100100100100100100
+                        submodule.qzeros.data[:,range(1,submodule.qzeros.data.shape[1],3)] += 0b10010010010010010010010010010010
+                        submodule.qzeros.data[:,range(2,submodule.qzeros.data.shape[1],3)] += 0b01001001001001001001001001001001
+                    elif quantize_config.bits == 4:
+                        submodule.qzeros.data += 0b00010001000100010001000100010001
+                    elif quantize_config.bits == 8:
+                        submodule.qzeros.data += 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+            else:
+                if use_qigen:
+                    submodule.zeros.data -= 1
+                else:
+                    if quantize_config.bits == 2:
+                        submodule.qzeros.data -= 0b01010101010101010101010101010101
+                    elif quantize_config.bits == 3:
+                        submodule.qzeros.data[:,range(0,submodule.qzeros.data.shape[1],3)] -= 0b00100100100100100100100100100100
+                        submodule.qzeros.data[:,range(1,submodule.qzeros.data.shape[1],3)] -= 0b10010010010010010010010010010010
+                        submodule.qzeros.data[:,range(2,submodule.qzeros.data.shape[1],3)] -= 0b01001001001001001001001001001001
+                    elif quantize_config.bits == 4:
+                        submodule.qzeros.data -= 0b00010001000100010001000100010001
+                    elif quantize_config.bits == 8:
+                        submodule.qzeros.data -= 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+
+    return model
```
This [comment](https://github.com/AutoGPTQ/AutoGPTQ/pull/559#discussion_r1529901562) is not addressed yet

> This does not check for overflows, this is an issue. We used to check for overflows in the kernels.

> 11111 will become 0000.

As I recall, this was an issue with AWQ checkpoints.

fxmarty replied:
Actually this may be fine

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -526,81 +570,101 @@ def save_quantized(
         self.quantize_config.model_file_base_name = model_base_name
 
+        # revert model
+        if revert_model is not None:
+            self.model = revert_model
```
The modif in `convert_gptq_v1_to_v2_format` is done in place, so this won't work.

fxmarty commented:
@Qubitium I'll push some modifs if that's fine to you

fxmarty commented on auto_gptq/modeling/_utils.py:
```diff
@@ -146,6 +147,51 @@ def make_quant(
+                        submodule.qzeros.data += 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+            else:
+                if use_qigen:
+                    submodule.zeros.data -= 1
+                else:
+                    if quantize_config.bits == 2:
+                        submodule.qzeros.data -= 0b01010101010101010101010101010101
+                    elif quantize_config.bits == 3:
+                        submodule.qzeros.data[:,range(0,submodule.qzeros.data.shape[1],3)] -= 0b00100100100100100100100100100100
+                        submodule.qzeros.data[:,range(1,submodule.qzeros.data.shape[1],3)] -= 0b10010010010010010010010010010010
+                        submodule.qzeros.data[:,range(2,submodule.qzeros.data.shape[1],3)] -= 0b01001001001001001001001001001001
+                    elif quantize_config.bits == 4:
+                        submodule.qzeros.data -= 0b00010001000100010001000100010001
+                    elif quantize_config.bits == 8:
+                        submodule.qzeros.data -= 0b00000001000000010000000100000001
+                    else:
+                        raise NotImplementedError("Only 2,3,4,8 bits are supported.")
```
The real issue is v2 -> v1 that may underflow

The underflow is not addressed in the AutoGPTQ codebase https://github.com/AutoGPTQ/AutoGPTQ/blob/ea829c7bbe83561c2b1de26795b6592992373ef7/auto_gptq/nn_modules/qlinear/qlinear_cuda_old.py#L166 anyway, so I'm not sure

Qubitium replied:
@qwopqwop200  Can you comment on this underflow note as whether if it is an issue or not? Feel like this is the last blocking discussion.

qwopqwop200 replied:
Almost all GPTQ checkpoints that exist today have sym=True (otherwise they won't work).
And the underflow issue is not a problem, at least when sym=True.
This is because when sym is True, the value of qzeros is fixed to 2** (bit-1).
Therefore, convert_gptq_v2_to_v1_format should only be applied to models converted from v1 to v2 or when sym=True.

Qubitium commented:
> @Qubitium I'll push some modifs if that's fine to you

Feel free to push changes as you see fit.

Qubitium commented:
@fxmarty My last todo (vllm/sglang compat) just completed and there is a compat problem with third party inference engine such as vllm. Maybe prudent to change to default to save in gptq (v1) format. Also, I have not yet tested saving sym=False to gptq v1 format and checking its compat with vllm.

Qubitium commented on auto_gptq/modeling/_base.py:
```diff
@@ -370,28 +368,10 @@ def tmp(_, inp, out):
-                        stat = {"layer": i + 1, "module": name, "avg_loss": avg_loss, "time": duration}
+                        stat = {"layer": i, "module": name, "avg_loss": avg_loss, "time": duration}
 
-                        # generate metric diff
```
@fxmarty I believe the grid printing (tabulate) and coloring of diff (termcolor) dependencies caused this part of the to be removed. Is there a way to work around this? This is very important to save time in quant validation/testing. gptq quanting is very sensitive to params (gpsize, dam, nsamples) etc and extremely long. It can take several hours to complete a quant. Now, users/tests can easier eye-ball the avg_loss as the logs stream and do a in-brain diff, or wait for quant to complete to do PPL. Both are doable but inefficient. Having a runtime diff of a saved baseline (quant-log) anyone can tell the new quant, due to whatever param change, is going to be bad in only few early layers and early stop quant.

fxmarty replied:
We could just add the ability to pass previous logs, and print the diff with it. I would just like to avoid to add dependencies.

Is this fine to do in an other PR? This one is already huge.

Qubitium commented on auto_gptq/modeling/_base.py:
```diff
@@ -370,28 +368,10 @@ def tmp(_, inp, out):
                         )
 
-                        stat = {"layer": i + 1, "module": name, "avg_loss": avg_loss, "time": duration}
+                        stat = {"layer": i, "module": name, "avg_loss": avg_loss, "time": duration}
```
@fxmarty  Sorry if duplicate. I can't find the original msg for some reason.  Can we keep `i + 1` index notation? This stat line is printed at each layer/module and at the very end (summary) but the logs are interjected with tqdm progress logs as well and they using `i + 1` index.  This stat is meant for visual consumption and having it be consistent with tqdm make the logs less confusing to the eye.

fxmarty replied:
yes yes let's keep it - should have deleted my comment instead of closing

Qubitium commented on auto_gptq/modeling/_base.py:
```diff
@@ -187,22 +193,17 @@ def quantize(
-----
-"""
-        logger.warning(thread_warning)
+        if "OPENBLAS_NUM_THREADS" not in os.environ or int(os.environ["OPENBLAS_NUM_THREADS"]) > 8:
+            thread_warning = """If you have not already done so, please inject the following code at the very top of your
+    quantization script so the packing stage is optimized for speed. Using too many cores may reduce packing performance.
+    ----
+    import os
+    os.environ['OPENBLAS_NUM_THREADS'] = str(1)
+    ----
+    """
+            logger.warning(thread_warning)
```
@fxmarty  We found issues with this change. Testing right now and should have more correct fix in coming hours.

Qubitium replied:
should be fixed with https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/a3863a44e246340496d0313b8c57324338e8993d  but need further testing.

Qubitium commented on auto_gptq/__init__.py:
```diff
@@ -1,6 +1,4 @@
-
-
-__version__ = "0.8.0.dev0"
+from .version import __version__
```
@fxmarty  Can you check if there is a missing file `version` push on your end? This line and __base.py with same new version code is breaking pkg.

fxmarty replied:
yes it was likely missing my bad

Qubitium commented:
Lots of users user autogptq and only a quant tool and not an inference engine so we need be careful about vllm like compat. Right now `gptq_v2` failed so likely kernel mod is required there too. `gptq` (v1) with `sym=False`  test pending..

* [x] Failed: Check if third party vllm/sglang needs kernel modification for the new `gptq_v2` format
* [ ] Check if third party vllm/sglang needs kernel modification for the new `gptq` (v1) format using `sym=False` in this PR

Qubitium commented:
Switching to draft mode so it doesn't get accidentally merged while I push some changes regarding packing (thread regression) for my team to test.

fxmarty commented:
@Qubitium Do you think we should by default save models in the old `gptq` format?

Qubitium commented:
> @Qubitium Do you think we should by default save models in the old `gptq` format?

Actually my team is testing this right now.  If gptq_v2 to gptq (v1) conversion has no issue and plays well with vllm, then we should default save to gptq (v1) for compat since almost no inference lib will support v2 without some kind of kernel mod.

Qubitium commented on auto_gptq/modeling/_utils.py:
```diff
@@ -297,32 +354,37 @@ def pack_model(
-        else:
-            qlayers[name].pack(layers[name], scale, zero, g_idx)
-        qlayers[name].to(layer_device)
+    # TODO remove once pack() thread regression is found
+    # Limit numpy thread usage in pack avoid slowdown: user_api=None to set limit to all supported libs
+    with tctl.threadpool_limits(limits=4, user_api=None):
+        pbar = tqdm(qlayers.keys(), leave=True)
+        for name in pbar:
+            pbar.set_description(f"Packing {name}")
+
+            quantizers[name], scale, zero, g_idx = quantizers[name]
+            # so far can only pack layer on CPU
+            layer_device = qlayers[name].device
+            qlayers[name].to(CPU)
+            layers[name], scale, zero, g_idx = (
+                layers[name].to(CPU),
+                scale.to(CPU),
+                zero.to(CPU),
+                g_idx.to(CPU),
+            )
+            if QuantLinear.QUANT_TYPE == "marlin":
+                qlayers[name].pack(layers[name], scale)
+            else:
+                qlayers[name].pack(layers[name], scale, zero, g_idx)
+            qlayers[name].to(layer_device)
+
```
@fxmarty  I know you don't want more depends but I found a way to fix the pack thread regression via code but need the `threadpoolctl` pkg.  I have confirmed the fix and we can get rid of nasty os.env injection and warning logs. 

https://github.com/joblib/threadpoolctl

Qubitium replied:
I will push this as separate PR since this PR may need a lot more testing (time) than I had initially thought.

Qubitium commented:
@fxmarty  @qwopqwop200 

Here is our test result. Based on this, we may need to consider the option to never expose `gptq_v2` as savable format and only use it as internal format with dynamic/runtime convert. So always save to `gptq (v1)` or at the very least have `gptq (v1)` as default save since the benefit for saving to `gptq_v2` is minimal and applies only if you also inference with autogptq tool itself. 

Note on columns:

* Quant is quantized model with `sym` value and Autogptq version that made the quant. 
* Next 3 is inference engine running the quant. 
* NO means unsupported or inference coherence is unusable. 
* YES means load ok and inference quality compared to baseline is normal.
* +YES means closest to baseline.

```
| Quant                     | AutoGPTQ Main | AutoGPTQ PR640 | VLLM (main) |
| :------------------------ | ------------: | :------------: | :---------: |
| Sym=True gptq (v1) Main   |           YES |      YES       |     YES     |
| Sym=True gptq (v1) PR640  |           YES |      YES       |     YES     |
| Sym=False gptq (v1) PR640 |          +YES |     +YES       |    +YES     |
| Sym=True gptq_v2 PR640    |            NO |      YES       |      NO     |
| Sym=False gptq_v2 PR640   |            NO |     +YES       |      NO     |
```

We had baseline of non-quant and compared to baseline, `sym=False` was the closest to baseline which is good. Notable is that `sym=False` in this PR but saved to `gptq (v1)` also runs on the main branch and vllm with same inference quality/output. 

`sym=False` saved to `gptq (v1)` using this PR retrains same quality and backward compatible with autogptq main and third party vllm/sglang. 

So our question becomes, why even bother saving to `gptq_v2` on disk? Only benefit I see for now is slightly faster loading time but incompatible with older autogptq and all 3rd gptq based inference libs. 

My vote is to keep `gptq_v2` an internal native format. 

Disclaimer: model tested is `TinyLlama-1.1B-intermediate-step-1341k-3T` and we only used one example prompt to test inference output quality. Different models, quants, prompts results may vary.

fxmarty commented:
@Qubitium yes, the issue is maybe underflow in https://github.com/Qubitium/AutoGPTQ/blob/9d88639600d263e46cec73fe5d03cc609bf3705b/auto_gptq/modeling/_utils.py#L192-L201

qwopqwop200 commented:
If sym=False you should force it to be saved in v2 format. It won't work in v1.
https://github.com/AutoGPTQ/AutoGPTQ/pull/559/files#diff-f4f987d3fa40acd4e45fc6fcd85d17228a61466910d12dfec21c918674979bacR105

fxmarty commented:
The thing is that e.g. vllm kernels have not been updated accordingly & won't load v2 models.

Qubitium commented:
> If sym=False you should force it to be saved in v2 format. It won't work in v1. https://github.com/AutoGPTQ/AutoGPTQ/pull/559/files#diff-f4f987d3fa40acd4e45fc6fcd85d17228a61466910d12dfec21c918674979bacR105

On `TinyLlama-1.1B-intermediate-step-1341k-3T` the underflow percentage of all qzero.data tensor elements is `~13%` for `convert_v2_to_v1()` with `sym=False`. 

@qwopqwop200 During my tests, despite the 13% underflows, the inference output generate is still good to great with forced `convert_v2_to_v1()` for `sym=False`. In theory, this should never work right? But can we assume the errors (underflows) are not affecting the model execution as expected?

Qubitium commented:
Failed Test
* [x] Yi9B with `sym=False` and `gptq_v2`. This bug is caused by #614

Qubitium commented on auto_gptq/modeling/_utils.py:
```diff
@@ -196,12 +199,57 @@ def convert_gptq_v2_to_v1_format(
+                    #             t[i] = subtract_result
+                    #         else:
+                    #             underflow += 1
+                    #             print(f"Underflow detected. Value not subtracted. module: {submodule}, tensor index: {i}, value:{val}")
```
@fxmarty @qwopqwop200  Please check if my underflow check code is accurate there. I pushed this to assist with testing for now. Thanks.

Qubitium commented:
@qwopqwop200  @fxmarty  Added underflow check for testing in https://github.com/AutoGPTQ/AutoGPTQ/pull/640/files#diff-426b34df9e6c820d8f15217a86004977116ade694a921b73edb49a4023622b30R201-R236

With `TinyLlama-1.1B-intermediate-step-1341k-3T` and `sym=Fase` and saved to `gptq (v1)` via `convert_v2_to_v1()`. The inference result is coherent and same as pre-quant model. However, if tweak the commit and skip the underflow math for tensor.items that underflow, the result is incoherent. So underlow is better than skip.

### TinyLlama-1.1B-intermediate-step-1341k-3T

### original model (pre-quant)
```
-- prompt: <s> My name is Lewis and I like to
-- output: <s> My name is Lewis and I like to play football.
I'm a 16 year old boy who loves playing football, but also enjoys other sports such as basketball, tennis etc.</s>


-- prompt: <s> Once upon a time,
<s> Once upon a time, there was an old man who lived in the forest. He had two sons and one daughter. The oldest son married a beautiful princess named Princess Ariel. They were happy together for many years until they decided to have children of their own. One day, when he went out hunting with his friends, the youngest son found himself alone in the woods.
The youngest son's name is Peter Pan. He has magical powers that allow him to fly through the air at will. Whenever he flies away from home, no matter how far it may be, he always comes back safely. This story tells us about
```

### quant: `sym=False` + save to `gptq (v1)` via `convert_v2_to_v1()` code path + underflowable qzero.data.items math skipped
```
-- prompt: <s> My name is Lewis and I like to
-- output: <s> My name is Lewis and I like toillet Comunagybraslide Kommunvez коммуandribusbjectLENGsegu↔ ÉsentovosеньommunÅilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletillet

-- prompt: <s> Once upon a time,
-- output: <s> Once upon a time,illet Comunagybraslide Kommunvez коммуandribusbjectLENGsegu↔ ÉsentovosеньommunÅilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletilletillet
```

### quant: `sym=False` + save to `gptq (v1)` via `convert_v2_to_v1()` code path + do all math even if underflow

```
-- prompt: <s> My name is Lewis and I like to
<s> My name is Lewis and I like to play football.
My favorite foods are pizza, chicken nuggets, and hotdogs.</s>

-- prompt: <s> Once upon a time,
<s> Once upon a time, there was an old man who lived in the forest. He had no family and he didn't have any friends either. One day, his friend came to visit him. The old man said that he would be happy if they could meet again one day.
The next morning, when the old man woke up, he found himself surrounded by flowers. They were so beautiful that it made him feel very sad. "What are you doing here?" asked the old man as he looked at them.
"I am your flower," replied the flowers. "You can come back whenever you want."
That night, the old man
```

Qubitium commented:
@fxmarty  At a fork (figurative) right now. Need to decide how to enforce this underflow safety check. My tests shows that underflow is not turning model is unusable. In fact, `sym=False` is good with or without underflow (gptq v1 underflow vs gptq_v2) and better than `sym=True`. 

* Should we outright ban saving `sym=False` as `gptq v1` as @qwopqwop200 suggested since math wise, this doesn't make sense if underflow happens. 
* Or allow with huge warnings?
* Or allow only with manual override in `.quantize(allow_unsafe_math=True)`?

There are still some testing I would like to do but personally, based on what I have seen, I would still want to use/allow `sym=False` and save to `gptq (v1)`.

Qubitium commented:
```
# yi9b
Underflow Summary: underflow count 973756, 12.006737294823232%
```

```
# command-r-v01
Underflow Summary: underflow count 3897083, 12.135666827766263%
```

Qubitium commented:
```
# llama3-8B

Underflow Summary: underflow count 875563, 12.846183776855469%
```

So tested 4 models and all points to a trend of underflow prob of <= 13% when doing `sym=False` and saving to gptq v1 via unsafe math convert_from_v2_to_v1() code path.

Qubitium commented:
@fxmarty  @qwopqwop200 

I do not understand why the <=13% underflow/overflow is not corrupting the model to an extend that make it the model unusable. In fact, my testing with 4 different models (tinyllama 1.1b, yi9b, command-r-v01, llama3-8b) shows that the underflow/overflow still result in stable generation with stable ppl that is onpar or better than sym=True with zero under/overflow saving to `gptq v1`.

Commit: https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/005a9422b78d326db51761a0fd4f89a75fe262d1 disable unsafe math (under/overflow) operations: 1) sym=False loading from gptq v1 and 2) sym=False saving to gptq v1.  But I also added `use_unsafe_math` argument to `from_quantized()` and `save_quantized()` to allow such unsafe ops for advanced users who understand the ramifications and risks involved.

qwopqwop200 commented:
I can't validate this idea at the moment as huggingface is throwing a 504.
It's easy to verify: check if the weights of qzeros in the v2 model and the v2 -> v1 -> v2 model are different.
Underflow may not be a problem. In C++, I simply underflowed an int32 and re-added it, and found that it returned to the normal weight value. 
However, I suspect that negative values could potentially be an issue, which could be solved by using uint32.

qwopqwop200 commented:
```
from transformers import AutoTokenizer, TextGenerationPipeline
from auto_gptq import AutoGPTQForCausalLM, BaseQuantizeConfig
import logging

logging.basicConfig(
    format="%(asctime)s %(levelname)s [%(name)s] %(message)s", level=logging.INFO, datefmt="%Y-%m-%d %H:%M:%S"
)

pretrained_model_dir = "facebook/opt-125m"
quantized_model_dir = "opt-125m-4bit"

tokenizer = AutoTokenizer.from_pretrained(pretrained_model_dir, use_fast=True)
examples = [
    tokenizer(
        "auto-gptq is an easy-to-use model quantization library with user-friendly apis, based on GPTQ algorithm."
    )
]

quantize_config = BaseQuantizeConfig(
    bits=4,  # quantize model to 4-bit
    group_size=128,  # it is recommended to set the value to 128
    sym=False,
    desc_act=False,  # set to False can significantly speed up inference but the perplexity may slightly bad
    checkpoint_format='gptq'
)

# load un-quantized model, by default, the model will always be loaded into CPU memory
model = AutoGPTQForCausalLM.from_pretrained(pretrained_model_dir, quantize_config,resume_download=True)

# quantize model, the examples should be list of dict whose keys can only be "input_ids" and "attention_mask"
model.quantize(examples)

# save quantized model
model.save_quantized(quantized_model_dir, use_unsafe_math=True)

# load quantized model to the first GPU
model2 = AutoGPTQForCausalLM.from_quantized('./opt-125m-4bit/', device="cuda:0")

assert id(model) != id(model2), 'this model is same'

for i in range(len(model.model.model.decoder.layers)):
    z0 = model.model.model.decoder.layers[i].self_attn.k_proj.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].self_attn.k_proj.qzeros.to('cpu')
    print(z0[z0 != z1])
    z0 = model.model.model.decoder.layers[i].self_attn.v_proj.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].self_attn.v_proj.qzeros.to('cpu')
    print(z0[z0 != z1])
    z0 = model.model.model.decoder.layers[i].self_attn.q_proj.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].self_attn.q_proj.qzeros.to('cpu')
    print(z0[z0 != z1])
    z0 = model.model.model.decoder.layers[i].self_attn.out_proj.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].self_attn.out_proj.qzeros.to('cpu')
    print(z0[z0 != z1])
    z0 = model.model.model.decoder.layers[i].fc1.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].fc1.qzeros.to('cpu')
    print(z0[z0 != z1])
    z0 = model.model.model.decoder.layers[i].fc2.qzeros.to('cpu')
    z1 = model2.model.model.decoder.layers[i].fc2.qzeros.to('cpu')
    print(z0[z0 != z1])
```

Qubitium commented:
> Underflow may not be a problem. In C++, I simply underflowed an int32 and re-added it, and found that it returned to the normal weight value. However, I suspect that negative values could potentially be an issue, which could be solved by using uint32.

There is small issue that we can't distinguish `sym=False` v1 from pre-PR code and v1 from this PR forward but I think we can safely ignore since `sym=False` never worked pre-PR so we are not breaking compat. We can just output a warning that v1 loading with `sym=False` is only safely supported if the original quant is using AutoGPTQ >= 0.8.x.

qwopqwop200 commented:
I have verified that both overflow and underflow issues do not occur. 
Here's the test code.
```
import torch
values = {'2': 0b01010101010101010101010101010101,
 '3-1': 0b00100100100100100100100100100100,
 '3-2': 0b10010010010010010010010010010010,
 '3-3': 0b01001001001001001001001001001001,
 '4': 0b00010001000100010001000100010001,
 '8': 0b00000001000000010000000100000001,}
int_range = (-2147483648, 2147483647)

for name, value in values.items():
    a = torch.arange(int_range[0], int_range[1] + 1, dtype=torch.int32)
    b = (a - value) + value
    print(name, 'underflow')
    print(a[a != b])
    del a,b

    a = torch.arange(int_range[0], int_range[1] + 1, dtype=torch.int32)
    b = (a + value) - value
    print(name, 'overflow')
    print(a[a != b])
    del a,b
```

Qubitium commented:
Below is PPL of a private model using command-r-v01. We tested ppl using vllm to show that even with the uncorrected underflow of v2 -> v1 for sym=False, the ppl/generation is very stable. @qwopqwop200 Showed the the underflow in save will be autocorrected/negated with overflow in autogptq loading.  In our testing, even without the auto-correction, saved gptq (v1) models with `sym=False` is quite good and offers full compat with 3rd party libs. 

### model: command-r-v01-trained

### bfloat16
```
vllm ppl: 1.6494
transformers ppl: 1.7075
``` 
 
### gptq(v1), sym=False
```
vllm ppl: 1.6515
autogptq ppl: 1.7084
```
 
### marlin, sym=True
```
vllm load: 1.6541,
autogptq load: 1.7139
```

Qubitium commented:
> Failed Test
> 
> * [x]  Yi9B with `sym=False` and `gptq_v2`. This bug is caused not by this PR changes but [[BUG] Regression in quantized inference when paired with Transformers >= 4.39.0 #614](https://github.com/AutoGPTQ/AutoGPTQ/issues/614)

This test now passing due to PR fix https://github.com/huggingface/transformers/pull/30380

Qubitium commented:
@qwopqwop200  @fxmarty 

To summarize the underflow/overflow situation.  Please correct me if anything is wrong here. 

1. If saving `sym=False` using autogptq, v2 -> v1 in save, then v1 -> v2 on load also using autogptq, this path is stable and safe. 
2. If saving `sym=False` using autogptq v2->1 in save, then load via 3rd party lib such as vllm, my tests show this is also stable (for all 4 diff models I have tried) though I cannot explain why it works despite the uncorrected underflows. 
3. Loading `sym=False` from quanted models pre-PR behavior is undefined and we currently do not have a way to check if v1 model is from pre-PR or post-PR (need another meta/info variable like tool/tool_version in quantized_config.json).

qwopqwop200 commented:
Yes. It would be nice to have some additional information in the config, like tool version or format version.
and 
I think this code should be cleaned up later, as the underflow issue seems to be fixed for now.
https://github.com/AutoGPTQ/AutoGPTQ/pull/640/files#diff-426b34df9e6c820d8f15217a86004977116ade694a921b73edb49a4023622b30R203-R258

Qubitium commented:
> I think this code should be cleaned up later, as the underflow issue seems to be fixed for now. https://github.com/AutoGPTQ/AutoGPTQ/pull/640/files#diff-426b34df9e6c820d8f15217a86004977116ade694a921b73edb49a4023622b30R203-R258

Underflow stat code removed now that we have done all the testing and know what to expect.

Qubitium commented:
Convert v1/v2 now faster with fix at https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/ac63033eec3effc1c33eb2e283541b1231708917

test_q4 now completes 40% faster on my zen3 dev env. 1000s vs 1650s. 

This is the same thread limit used on pack(): https://github.com/AutoGPTQ/AutoGPTQ/pull/642 Whatever auto-penalization torch/cublas is doing should be disabled.

Qubitium commented:
@fxmarty  Read to re-review. 

Even though v2->v1 then v1->v2 is validated safe, we can't be sure quant model will be loaded/inferenced with autogptq. User may quant v2 -> v1 and then use the v1 model on vllm or another. As such, the `use_unsafe_math` parameter is still needed to cautious. 

Loading behavior of v1 pre-pr quanted with `sym=False` is undefined. We can only disable this code path with extra meta fields in `quantized_config.json` but I think it should be in another pr.

Qubitium commented:
@qwopqwop200  @fxmarty  I am going to inject new meta code into this PR to complete the final missing piece of preventing the auto convertion of v1 to v2 when loading models not quantized by pre-PR code.  The new code is minimal. https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/a6d64e99756e5fd9ff1753c6595a25096f4af9d6

```
# sample new field meta (dic) with "quantizer" as the field holding autogptq fingerprint + version
meta={'quantizer': 'autogptq:0.8.1.dev0'}
```

There are many paths to add meta info. Here is my take:

* Use a `meta` dict to hold all new properties that is not directly involved the autogptq quantization and/or loading/inference of the model
* `meta.quantizer` field has format of `tool name:version` separted by `:` => "autogptq:0.8.1"
* Fields like timestamps, who made the quant, git commit hah of model, urls. etc.
* Anyone can add to it via `meta_set` and `meta_get` methods to `BasedQuantrizedConfig`

This way, we have a single meta property field that can be used us and others without polluting the top level vars. Intel's autoround (gptq) comes to mind as they inject lots of stuff into quantized_config.json with vars only related to their tooling that should go into meta.

Qubitium commented:
Fixed meta json save and added meta field validation to test_quantization

@fxmarty Another pkg depend added `packaging`: since we need a python native version string checker and `distutils` is deprecated in favor of `packaging.version`.

Qubitium commented:
Fixed awq pack/unpack thread regression in https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/cddbe23b17de64e16b7c06e674f23f089dc57dbc

unpack/packing to awq in test_awq now ~2.6x faster. 2m12s vs 5m46s.

Qubitium commented:
All tests passing with transformers 4.40.1

* [x] test_q4
* [x] test_quantization
* [x] test_serialization  
* [x] test_shareded_load
* [x] test_awq_compat

Qubitium commented:
Remove highly flaky Mixtral-Tiny model, that doesn't exist in any production environment, and test that 1) fails at a very high rate 2) input/output are nonsensical and no human can understand what this is testing. Only the model creator (hf) knows what inside this black genie bottle in terms of weights. 

https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/811ca399f8c770143640c624ce115f7b32d800d6

Qubitium commented:
Removed all `use_unsafe_math` code and default save to `gptq (v1)`format for max compat with minimal loss vs saving to v2 as tested in our internal ppl tests across 4 different models. 

@fxmarty  Feel free to edit at will and revert the last commit if you want to go the conservative route. https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/528a8fce5aa92917db6b7f9a772eec68f67b2d70

Qubitium commented:
Intel/auto-round by default uses `sym=False` and save to gptq (v1) format by importing autogptq lib and directly calling `pack()` . As it stands, we are going to reject loading this model. I need to test this. My lord, testing for this PR never ends. I am going to pull my hair out. 

```json
{
  "bits": 4,
  "group_size": 128,
  "damp_percent": 0.01,
  "desc_act": false,
  "static_groups": false,
  "sym": false,
  "true_sequential": false,
  "model_name_or_path": null,
  "model_file_base_name": "model",
  "quant_method": "gptq",
  "checkpoint_format": "gptq",
  "meta": {
    "quantizer": "intel/auto-round:0.1",
    "iters": 10,
    "lr": 0.1,
    "minmax_lr": 0.1,
    "enable_minmax_tuning": true,
    "use_quant_input": true,
    "scale_dtype": "torch.float16"
  }

}
```

Qubitium commented:
> Intel/auto-round by default uses `sym=False` and save to gptq (v1) format by importing autogptq lib and directly calling `pack()` . As it stands, we are going to reject loading this model. I need to test this. 

* [x] Autoround compat confirmed with Intel/auto-round compatibility confirmed. Needs this https://github.com/intel/auto-round/pull/87 and tested with both sym=True/False.  

Add `meta.packer` so it is clear who is doing what. Autoround is doing the quantization and autogptq (imported code) is doing the packing to v1.

autoround model config with sym=False
```json
{
  "bits": 4,
  "group_size": 128,
  "damp_percent": 0.01,
  "desc_act": false,
  "static_groups": false,
  "sym": false,
  "true_sequential": false,
  "model_name_or_path": null,
  "model_file_base_name": "model",
  "quant_method": "gptq",
  "checkpoint_format": "gptq",
  "meta": {
    "quantizer": "intel/auto-round:0.1",
    "packer": "autogptq:0.8.0.dev1",
    "iters": 20,
    "lr": 0.05,
    "minmax_lr": 0.05,
    "enable_minmax_tuning": true,
    "use_quant_input": true,
    "scale_dtype": "torch.float16"
  }
}
````

Qubitium commented:
Commit/Refractor https://github.com/AutoGPTQ/AutoGPTQ/pull/640/commits/59be4b35b09fa42a7fab9ede5d5cad9c4dce199c

Tooling is now broken down to  `meta.quantizer` and `meta.packer` as evident by intel/auto-round usage of autogptq packer but quantization is done by auto-round. Medels quanted by autogptq does not set `meta.packer`.  v1 checks both `meta.quantizer` and `meta.packer` to see if it was either quantized or packed by autogptq. 

```
 "meta": {
    "quantizer": "intel/auto-round:0.1",
    "packer": "autogptq:0.8.0.dev1",
}
```

wenhuach21 commented:
We found a significant discrepancy between the real quantized model and the QDQ model at W2G32 asym. However, for the sym and W4G128 asym, there doesn't seem to be severe issue.

As this PR also supports the v2 format and we are not familiar with the details, we're unsure if we can test it directly or if we need to merge the PR in Autoround.

wenhuach21 commented:
This is our the results, we used autoround to generate the qdq model and real autogptq model by "--deployment_device fake,gpu", then we could evaluate the both model. lm-eval 0.4.2 is used
asym at w2g32
qdq:   lambada_openai: 0.5405   winogrande: 0.6417
gpu:   lambada_openai: 0.2505   winogrande: 0.6006

sym at w2g32
qdq:   lambada_openai: 0.3501  winogrande: 0.5817
gpu:   lambada_openai: 0.3629   winogrande: 0.5856


reference command
git clone https://github.com/intel/auto-round.git
cd auto-round/examples/language-modeling
pip install -r requirements.txt

python3 main.py --model_name /data5/llama3_8b_instruct/ --bits 2 --group_size 32 --n_samples 512 --iters 200 --deployment_device fake,gpu --disable_eval --seqlen 512 --minmax_lr 0.01 --scale_dtype fp32  --train_bs 4 --output_dir "./tmp_signround"

wenhuach21 commented:
we also find large discrepancy for phi-2 at W4G128 asym. Shall I create a new issue for this?

Qubitium commented:
@qwopqwop200  For various reasons my team has forked and refractored AutoGPTQ into new [GPTQModel](https://github.com/ModelCloud/GPTQModel) project. This bug fix/feature PR has been merged there. I have done plenty of testing and validation for this PR and it works, not perfectly but it works. Default save to v1 for max compat.

RanchiZhao commented:
> @qwopqwop200 This is the rebase of your PR at #559 with some modifications. Should be ready soon after we verify quantize, inference, and add some tests. 这是您在 #559 处的 PR 的变基，并进行了一些修改。在我们验证量化、推理并添加一些测试后应该很快就可以准备好。
> 
> ### Reason For PR: 公关理由：
> `sym=False` was practically unusable due to post-quantization avg_loss per layer/PPL vs `sym=True` . @qwopqwop200 fixed the bad/suboptimal math. Now `sym=False` will most likely match or decrease avg_loss/layer vs `sym=True` and improve post-quant PPL for many models.由于每层后量化 avg_loss/PPL 与 `sym=True` 相比， `sym=False` 实际上无法使用。修复了错误/次优的数学问题。现在， `sym=False` 很可能会与 `sym=True` 相比匹配或减少 avg_loss/layer，并提高许多模型的量化后 PPL。
> 
> ### Core Changes: 核心变化：
> 1. Rebase PR [remove (zeors -= 1) #559](https://github.com/AutoGPTQ/AutoGPTQ/pull/559) with `main`: allow usable `sym=False` quantization and use `checkpoint_format=gptq_v2` to store new checkpoint format. Compat runtime dynamic convert of all `checkpoint_format=gptq` to `gptq_v2` on load.使用 `main` 重新设置 PR [remove (zeors -= 1) #559](https://github.com/AutoGPTQ/AutoGPTQ/pull/559) ：允许可用的 `sym=False` 量化并使用 `checkpoint_format=gptq_v2` 存储新的检查点格式。 Compat 运行时在加载时将所有 `checkpoint_format=gptq` 动态转换为 `gptq_v2` 。
> 
> ### Misc Changes not directly related to sym=False code:
> 其他与 sym=False 代码不直接相关的更改：
> 4. Complete Todo: Use accelerate 0.29.2 to load checkpoint.完成Todo：使用accelerate 0.29.2加载检查点。
> 5. Consistency: Move cohere/starcoder2 to 4.39.0 release check一致性：将 cohere/starcoder2 移至 4.39.0 版本检查
> 6. Usability: Catch and alert user on how to fix quant/torch error caused by low damp/nsamples [[BUG] torch._C._LinAlgError: linalg.cholesky always raised #572](https://github.com/AutoGPTQ/AutoGPTQ/issues/572) as this happens much more frequently than I had expected. Ran into 2-3 instances of this error on multiple models during testing for this PR when using low `nsamples`  + low `damp=0.005` to speed up quants.可用性：捕获并提醒用户如何修复由低阻尼/n样本 [[BUG] torch._C._LinAlgError: linalg.cholesky always raised #572](https://github.com/AutoGPTQ/AutoGPTQ/issues/572) 引起的定量/火炬错误，因为这种情况发生的频率比我预期的要高得多。在测试此 PR 期间，使用低 `nsamples` + 低 `damp=0.005` 来加速量化时，在多个模型上遇到了 2-3 个此错误实例。
> 7. Simplify: optimized packing regression alert message/code to user简化：向用户优化打包回归警报消息/代码
> 8. Feature: Quant Stat Log 1/2:  store per layer quant stats (`layer #`, `module name`, `avg loss`, `duration`)  in dict/slice and return to user va `quant_log = model.quantize()`功能：定量统计日志 1/2：将每层定量统计信息（ `layer #` 、 `module name` 、 `avg loss` 、 `duration` ）存储在 dict/ 中切片并返回给用户 va `quant_log = model.quantize()`
> 9. Feature: Quant Stat Log 2/2:  pass saved quant_log to `quantize(quant_log=saved_quant_log)` to generate auto-avg_loss diff in progress. Sample diff output in later messages of this discussion.功能：Quant Stat Log 2/2：将保存的 quant_log 传递到 `quantize(quant_log=saved_quant_log)` 以生成正在进行的自动 avg_loss diff。本讨论后面的消息中的示例 diff 输出。
> 10. Usability: Use tqdm for layer loop so users can have an estimate of quant remaining time.可用性：使用 tqdm 进行层循环，以便用户可以估计定量剩余时间。
> 
> TODO: 去做：
> 
> * [x]  Add `sym=False` tests  添加 `sym=False` 测试
> * [x]  Validate and fix failing tests验证并修复失败的测试
> * [x]  Failed: Check if third party vllm/sglang needs kernel modification for the new `gptq_v2` format 失败：检查第三方 vllm/sglang 是否需要内核修改以适应新的 `gptq_v2` 格式
> * [x]  Check if third party vllm/sglang needs kernel modification for the new `gptq` (v1) format using `sym=False` in this PR 检查第三方 vllm/sglang 是否需要使用此 PR 中的 `sym=False` 进行内核修改以适应新的 `gptq` (v1) 格式
> 
> PASSING TESTS: 通过测试：
> 
> * [x]  Compat: vllm/sglang compatible  with gptq (v1) generated with this PR Compat：vllm/sglang 与此 PR 生成的 gptq (v1) 兼容
> * [x]  Compat load of `checkpoint_format=gptq`(v1) `checkpoint_format=gptq` 的兼容负载 (v1)
> * [x]  `sym=False` consistently generate lower `avg_loss` than 始终生成低于`sym=True` `sym=False` 始终生成低于 `sym=True` 的 `avg_loss`
> * [x]  Regression Test: `sym=True` in PR generates same math/`avg_loss` for layers as `sym=True` in 回归测试：在 PR 中为各层生成相同的数学/`main` 回归测试：PR 中的 `sym=True` 为层生成与 `main` 中的 `sym=True` 相同的数学/ `avg_loss`
> * [x]  `test_serialization.py`
> * [x]  `test_quantization.py`
> * [x]  `test_shared_loading.py`
> * [x]  `test_awq_compatibility_generation.py`  (note: awq cache generated by `main` is not compatible with pr. fixed with version file name adding `v2`) `test_awq_compatibility_generation.py` （注意： `main` 生成的awq缓存与pr不兼容。已修复版本文件名添加 `v2` ）
> * [x]  `test_q4.py`
> 
> FAILING TESTS: 失败的测试：
> 
> * [x]  compat: vllm/sglang not compatible gptq_v2 兼容：vllm/sglang 不兼容 gptq_v2
> * [x]  `test_triton.py` (never got this to work on main) `test_triton.py` （从来没有让它在 main 上工作）
> * [x]  `test_repacking.py` (never got this to work on main) `test_repacking.py` （从来没有让它在 main 上工作）
> 
> Original PR #559 notes duplicated here for ref:原始 PR #559 注释复制于此以供参考：
> 
> > * [x]   check work marlin 检查工作马林鱼
> > * [x]   check work exllama 检查工作埃拉玛
> > * [x]   check work exllama2 检查工作 exllama2
> > * [x]   check work qigen 检查工作启根
> > * [x]   check work triton 检查工作 Triton
> > * [x]   check work cuda 检查工作 cuda
> > * [x]   ~check work cuda old~(There is a bug in autogptq's main unrelated to this pr, so it cannot be confirmed.)check work cuda old（autogptq的main中有一个与此pr无关的bug，因此无法确认。）
> > * [x]   check work cuda pytorch检查工作 cuda pytorch
> > * [x]   check work cuda old pytorch 检查 cuda 旧 pytorch 的工作
> > * [x]   check support old version save检查支持旧版本保存
> > * [x]   check support old version load 检查支持旧版本加载
> > * [x]   check support new version save 检查支持新版本保存
> > * [x]   check support new version load 检查支持新版本加载
> > 
> > I am removing this line because it is not only computationally unnecessary, but also makes sym=False impossible. However, it breaks backwards compatibility, so I making it the default to use the old save format.
> > ## related pr
> > #354
> > #325
> > I removed the following line: https://github.com/AutoGPTQ/AutoGPTQ/blob/main/auto_gptq/nn_modules/qlinear/qlinear_cuda.py#L169 This is an unnecessary line. And this line makes it impossible to sym='False' If sym='False' I get a reduction in ppl.
> > opt-125m(act-order)	Bits	group-size	Wikitext2
> > sym=True	4	128	29.875
> > sym=False	4	128	29.221
> > llama2(act-order)	Bits	group-size	Wikitext2
> > sym=True	4	128	5.254
> > sym=False	4	128	5.214

I wonder how to fix what you have said, specifically when i let attention_mask mask the insturctions of the calibration dataset?
"Usability: Catch and alert user on how to fix quant/torch error caused by low damp/nsamples https://github.com/AutoGPTQ/AutoGPTQ/issues/572 as this happens much more frequently than I had expected. Ran into 2-3 instances of this error on multiple models during testing for this PR when using low nsamples + low damp=0.005 to speed up quants."

this really annoy me
